From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Cox <jc@kynesim.co.uk>
Date: Thu, 23 Apr 2026 14:50:48 +0100
Subject: [PATCH 03/46] Copy start logic from FFmpeg to DRMPRIME

DRNPrime startup can sometimes produce mucky ouput. s/w startup
if OK. Copy the s/w logic to the DRMPrime logic.
As FFmpeg V4L2 m2m decode (at least on Pi) doesn't set frame type
and Key frame usefully (and at least on Pi it is tricky to make it
do so) treat picture type _NONE as good.

(cherry picked from commit a8abdbc858953842ccdc31c212511c4b6c832419)
---
 .../VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp     | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp b/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp
index c21451139f..eee23064e5 100644
--- a/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp
+++ b/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp
@@ -684,7 +684,7 @@ CDVDVideoCodec::VCReturn CDVDVideoCodecDRMPRIME::GetPicture(VideoPicture* pVideo
   int64_t framePTS = m_pFrame->best_effort_timestamp;
   m_dropCtrl.Process(framePTS, m_pCodecContext->skip_frame > AVDISCARD_DEFAULT);
 
-  if (m_pFrame->flags & AV_FRAME_FLAG_KEY)
+  if ((m_pFrame->flags & AV_FRAME_FLAG_KEY) != 0 || m_pFrame->pict_type == AV_PICTURE_TYPE_NONE)
   {
     m_started = true;
     m_iLastKeyframe = m_pCodecContext->has_b_frames + 2;
@@ -700,7 +700,7 @@ CDVDVideoCodec::VCReturn CDVDVideoCodecDRMPRIME::GetPicture(VideoPicture* pVideo
     int frames = 300;
     if (m_dropCtrl.m_state == CDropControl::VALID)
       frames = static_cast<int>(6000000 / m_dropCtrl.m_diffPTS);
-    if (m_iLastKeyframe >= frames && m_pFrame->pict_type == AV_PICTURE_TYPE_I)
+    if (m_iLastKeyframe >= frames && (m_pFrame->pict_type == AV_PICTURE_TYPE_I || m_pFrame->pict_type == AV_PICTURE_TYPE_NONE))
     {
       m_started = true;
     }
